File: Sep_23/23_9_19/sw_5209.py
import sys
sys.stdin = open('input.txt')


def find_ans(now, now_val):
    global res
    # 유망성 확인
    if res < now_val:
        return
    # 끝까지 도달하면 결과값 갱신
    if sum(visited) == n:
        res = min(now_val, res)
        return
    # 경우의수 나누기
    for i in range(n):
        if not visited[i]:
            visited[i] = 1
            find_ans(now + 1, now_val + v[now][i])
            visited[i] = 0


T = int(input())
for tc in range(1, T+1):
    n = int(input())
    v = [list(map(int, input().split())) for _ in range(n)]
    visited = [0] * n
    res = 1e9
    find_ans(0, 0)
    print(f'#{tc} {res}')

File: CPP/mac_boj/sw_5209.cpp
#include <iostream>
#include <vector>

using namespace std;

int res = 100000000;
int n;

void find_ans(int visited[], int now, int now_val, int arr[]){
    if (res < now_val) return ;

    int sum_val = 0;
    for (int i=0; i<n; i++){
        sum_val += visited[i];
    }
    if(sum_val == n){
        if (now_val < res) res = now_val;
        return ;
    }

    for(int i = 0; i<n; i++){
        if (visited[i] == 0){
            visited[i] = 1;
            find_ans(visited, now+1, now_val, arr);
            visited[i] = 0;
        }
    }
}

int main(){
    int T;
    cin >> T;
    for (int i=1; i<=T; i++){
        cin >> n;

        // 동적할당으로 입력
        // int** arr = new int*[n];
        // for(int i=0;i<n;i++){
        //     arr[i] = new int[n];
        //     fill_n(arr[i], n, 0);
        //     for (int j=0; j<n; j++){
        //         cin >> arr[i][j];
        //     }
        // }
        vector<vector<int>> arr(n);
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                int t;
                cin >> t;
                arr[i].push_back(t);
            }
        }
        vector<int> visited[n];
        // find_ans(visited, 0, 0, arr);
    }
}